File: 0377-combination-sum-iv/0377-combination-sum-iv.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var combinationSum4 = function(nums, target) {
  // Bottom Up - Tabulation method
  nums.sort((a,b) => a-b);
  const combos = new Array(target + 1).fill(0);
  // 1 combo to make target of 0...empty array []
  combos[0] = 1
  
  for (let t = 0; t < combos.length; t++) {
    for (let num of nums) {
      if (num > t) break;
      combos[t] += combos[t - num];
    }
  }
  
  return combos[target];  
  
  // Top Down - Memoization method
  // Base cases
//   if (target < 0) return 0; // no possible combos
//   if (target === 0) return 1; // only possible combo is []
//   if (target in memo) return memo[target];
  
//   let count = 0;
  
//   for (let num of nums) {
//     if (num <= target) {
//       const diff = target - num;
//       count += combinationSum4(nums, diff, memo);
//     }
//   }
  
//   memo[target] = count;
//   return count;
  
};

File: 0377-combination-sum-iv/0377-combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:

      def findCombos(target, memo):
        if target in memo:
          return memo[target]
        if target < 0: 
          return 0
        if target == 0:
          return 1
        
        count = 0
        for num in nums:
          if num <= target:
            diff = target - num
            count += findCombos(diff, memo)
        memo[target] = count
        return memo[target]
        
      return findCombos(target, {})
